In [29]:
#                                  OLS Regression Results                                
# =======================================================================================
# Dep. Variable:                results   R-squared (uncentered):                   0.316
# Model:                            OLS   Adj. R-squared (uncentered):              0.232
# Method:                 Least Squares   F-statistic:                              3.781
# Date:                Sun, 18 Sep 2022   Prob (F-statistic):                    9.88e-06
# Time:                        11:29:27   Log-Likelihood:                         -602.64
# No. Observations:                 147   AIC:                                      1237.
# Df Residuals:                     131   BIC:                                      1285.
# Df Model:                          16                                                  
# Covariance Type:            nonrobust                                                  
# ===================================================================================
#                       coef    std err          t      P>|t|      [0.025      0.975]


# Omnibus:                        3.493   Durbin-Watson:                   1.122
# Prob(Omnibus):                  0.174   Jarque-Bera (JB):                2.831
# Skew:                           0.223   Prob(JB):                        0.243
# Kurtosis:                       2.487   Cond. No.                         6.79

In [25]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_absolute_percentage_error

from sklearn import linear_model
import statsmodels.api as sm
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import statsmodels.api as sm

In [2]:
df = pd.read_csv('wc-wo-outliers.csv')

In [11]:
new_X = df.drop(columns=['results'])

In [12]:
new_X

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0    0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1    0.245385  0.097692   0.549231      0.490000  0.090769      0.071538   
2    0.337692  0.270000   0.292308      0.281538  0.065385     -0.142308   
3    0.673077  0.642308   0.502308      0.277692  0.310000      0.146923   
4    0.114615  0.155385   0.271538      0.330769 -0.015385      0.028462   
..        ...       ...        ...           ...       ...           ...   
179 -0.080769 -0.243846  -0.133846     -0.276154  0.096154      0.124615   
180 -0.066923 -0.147692   0.028462     -0.050000 -0.076154      0.314615   
181 -0.066154 -0.143846  -0.150769      0.256154  0.185385      0.366923   
182 -0.193846 -0.290769  -0.075385     -0.372308 -0.380000     -0.211538   
183 -0.044615 -0.314615  -0.078462     -0.232308 -0.153077     -0.076154   

     takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  \
0     0.244615   -0.220000   0.216154         0.270000     -0.076923   
1    -0.473077   -0.150769   0.096923         0.277692     -0.153077   
2    -0.076923    0.430000   0.688462         0.002308      0.020769   
3     0.153077    0.081538   0.814615         0.464615     -0.150769   
4     0.128462    0.109231   0.130769         0.056923     -0.232308   
..         ...         ...        ...              ...           ...   
179  -0.200769    0.102308   0.060000        -0.073077      0.062308   
180  -0.224615    0.122308  -0.207692         0.190769      0.413846   
181  -0.002308    0.240000   0.542308         0.206154      0.380769   
182  -0.340000   -0.636923  -0.611538        -0.588462      0.342308   
183   0.004615   -0.446154  -0.537692        -0.330000      0.531538   

     blocks_z  aerials_z   fouls_z  fouled_z    nsxg_z  
0   -0.097692  -0.020000 -0.224615  0.100769 -0.124615  
1   -0.212308   0.033846 -0.145385 -0.036154  0.487692  
2    0.165385  -0.075385  0.193077  0.087692  0.393846  
3   -0.007692  -0.060769 -0.220000  0.006154  0.387692  
4   -0.146923  -0.023077  0.078462  0.532308  0.140769  
..        ...        ...       ...       ...       ...  
179 -0.343077  -0.007692 -0.325385 -0.152308 -0.331538  
180 -0.044615  -0.035385 -0.173846 -0.323846 -0.006923  
181  0.153077   0.009231 -0.100769  0.133077 -0.080769  
182 -0.099231   0.266154 -0.360000 -0.209231 -0.310000  
183 -0.081538  -0.033077  0.741538  0.106923 -0.129231  

[184 rows x 16 columns]

In [13]:
y = df['results']

X_tr, X_tt, y_tr, y_tt = train_test_split(
     new_X, y, test_size=0.20, random_state=1)

In [14]:
y = df['results']
X = df.drop(columns=['results'])

X_train, X_test, y_train, y_test = train_test_split(
     X, y, test_size=0.20, random_state=1)

In [15]:
reg = linear_model.LinearRegression()
reg.fit(X_train, y_train)
pred_train= reg.predict(X_train)

In [16]:
reg = linear_model.LinearRegression()
reg.fit(X_tr, y_tr)
pred_new= reg.predict(X_tr)

Theres no clear difference between using all the features or a few features. mape goes back and forth depending on the random_state value

In [17]:
# print('Linear Regression ')
# print('---------------------')
# print('RMSE: '+str(np.sqrt(mean_squared_error(y_train,pred_train))))
# print('R Squared: '+str(r2_score(y_train, pred_train)))
# print('MSE: '+str(mean_squared_error(y_train, pred_train)))
# print('MAE: '+str(mean_absolute_error(y_train, pred_train)))
# print('MAPE: '+str(mean_absolute_percentage_error(y_train, pred_train)))

# print('---------------------')

pred_test= reg.predict(X_test)

print('RMSE : '+str(np.sqrt(mean_squared_error(y_test,pred_test)))) 
print('R2 : '+str(r2_score(y_test, pred_test)))
print('MSE: '+str(mean_squared_error(y_test, pred_test)))
print('MAE: '+str(mean_absolute_error(y_test, pred_test)))
print('MAPE: '+str(mean_absolute_percentage_error(y_test, pred_test)))

RMSE : 7.313588408701075
R2 : 0.32155561150656986
MSE: 53.48857541188672
MAE: 6.190975448358605
MAPE: 1.056767915202244


In [31]:
model = sm.OLS(y_train, X_train)
results = model.fit()
print(results.summary())

# https://www.statsmodels.org/dev/examples/notebooks/generated/ols.html


                                 OLS Regression Results                                
Dep. Variable:                results   R-squared (uncentered):                   0.316
Model:                            OLS   Adj. R-squared (uncentered):              0.232
Method:                 Least Squares   F-statistic:                              3.781
Date:                Sun, 18 Sep 2022   Prob (F-statistic):                    9.88e-06
Time:                        18:04:58   Log-Likelihood:                         -602.64
No. Observations:                 147   AIC:                                      1237.
Df Residuals:                     131   BIC:                                      1285.
Df Model:                          16                                                  
Covariance Type:            nonrobust                                                  
                      coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------

In [18]:
print('Linear Regression ')
print('---------------------')
print('RMSE: '+str(np.sqrt(mean_squared_error(y_tr,pred_new))))
print('R Squared: '+str(r2_score(y_tr, pred_new)))
print('MSE: '+str(mean_squared_error(y_tr, pred_new)))
print('MAE: '+str(mean_absolute_error(y_tr, pred_new)))
print('MAPE: '+str(mean_absolute_percentage_error(y_tr, pred_new)))

print('---------------------')

pred_tt= reg.predict(X_tt)

print('RMSE : '+str(np.sqrt(mean_squared_error(y_tt,pred_tt)))) 
print('R2 : '+str(r2_score(y_tt, pred_tt)))
print('MSE: '+str(mean_squared_error(y_tt, pred_tt)))
print('MAE: '+str(mean_absolute_error(y_tt, pred_tt)))
print('MAPE: '+str(mean_absolute_percentage_error(y_tt, pred_tt)))

Linear Regression 
---------------------
RMSE: 6.579236560582229
R Squared: 0.4575845582345789
MSE: 43.286353720101886
MAE: 5.445821401520943
MAPE: 0.8212018104777041
---------------------
RMSE : 7.313588408701075
R2 : 0.32155561150656986
MSE: 53.48857541188672
MAE: 6.190975448358605
MAPE: 1.056767915202244


In [19]:
x_train = X_train[['goals_z', 'xg_z', 'tackles_z', 'nsxg_z']]

In [20]:
x_test = X_test[['goals_z', 'xg_z', 'tackles_z', 'nsxg_z']]

In [21]:
reg.fit(x_train, y_train)
pred_train= reg.predict(x_train)

In [22]:
print('Linear Regression ')
print('---------------------')
print('RMSE: '+str(np.sqrt(mean_squared_error(y_train,pred_train))))
print('R Squared: '+str(r2_score(y_train, pred_train)))
print('MSE: '+str(mean_squared_error(y_train, pred_train)))
print('MAE: '+str(mean_absolute_error(y_train, pred_train)))
print('MAPE: '+str(mean_absolute_percentage_error(y_train, pred_train)))

print('---------------------')

pred_test= reg.predict(x_test)

print('RMSE : '+str(np.sqrt(mean_squared_error(y_test,pred_test)))) 
print('R2 : '+str(r2_score(y_test, pred_test)))
print('MSE: '+str(mean_squared_error(y_test, pred_test)))
print('MAE: '+str(mean_absolute_error(y_test, pred_test)))
print('MAPE: '+str(mean_absolute_percentage_error(y_test, pred_test)))

Linear Regression 
---------------------
RMSE: 6.7559220831846405
R Squared: 0.4280602185838347
MSE: 45.64248319406189
MAE: 5.630387090633862
MAPE: 0.8641609743558922
---------------------
RMSE : 7.387770211766454
R2 : 0.3077228757657464
MSE: 54.57914870186375
MAE: 6.33870065693538
MAPE: 1.1000763785952064


the lowest MAPE value, which tells us that it’s able to forecast the most accurately among the potential models.

In [10]:
from sklearn.metrics import mean_squared_log_error
print('Mean Squared Log Error: ' +str(mean_squared_log_error( y_test, pred_test, squared=False)))

Mean Squared Log Error: 0.5892030828713158


In [12]:
# #display adjusted R-squared
# print('Adjusted r^2: ' +str(1 - (1-reg.score(X_train, y_train))*(len(y_train)-1)/(len(y_train)-X_train.shape[1]-1)))

The value for R-squared can range from 0 to 1. A value of 0 indicates that the response variable cannot be explained by the predictor variable at all while a value of 1 indicates that the response variable can be perfectly explained without error by the predictor variables.

- AIC stands for (Akaike’s Information Criteria), a metric developped by the Japanese Statistician, Hirotugu Akaike, 1970. The basic idea of AIC is to penalize the inclusion of additional variables to a model. It adds a penalty that increases the error when including additional terms. The lower the AIC, the better the model.
- AICc is a version of AIC corrected for small sample sizes.
- BIC (or Bayesian information criteria) is a variant of AIC with a stronger penalty for including additional variables to the model.
- Mallows Cp: A variant of AIC developed by Colin Mallows.


# AIC

In [13]:
import statsmodels.api as sm

In [14]:
#add constant to predictor variables
X_train = sm.add_constant(X_train)

#fit regression model
model = sm.OLS(y_train, X_train).fit()

#view AIC of model
print('AIC: '+str(model.aic))

AIC: 804.0899081188967


In [15]:
X_train.head()

const   goals_z      xg_z  crosses_z  boxtouches_z  passes_z  \
126    1.0 -0.018462  0.184615   0.168462      0.276923 -0.106923   
97     1.0  0.198462 -0.116154  -0.090000     -0.013077  0.193077   
42     1.0  0.214615  0.523077  -0.176154      0.277692 -0.196923   
135    1.0  0.766923  0.643846  -0.265385      0.230000  0.353077   
28     1.0  0.332308  0.362308  -0.043846      0.316923  0.060000   

     progpasses_z  takeons_z  progruns_z  tackles_z  interceptions_z  \
126     -0.225385   0.138462    0.063846  -0.024615        -0.153846   
97       0.132308   0.010769    0.395385   0.102308         0.802308   
42      -0.245385  -0.013846   -0.178462  -0.228462        -0.250000   
135     -0.015385   0.496154    0.606923   0.446154        -0.096923   
28       0.019231   0.140769    0.210000   0.343846         0.276154   

     clearances_z  blocks_z  aerials_z   fouls_z  fouled_z    nsxg_z  
126     -0.119231  0.056923   0.285385 -0.036923 -0.293077  0.228462  
97      -0.154615 -0.013077   0.016154  0.668462  0.585385 -0.030000  
42       0.096154  0.032308   0.760769 -0.122308  0.119231  0.124615  
135     -0.128462  0.086923  -0.233846  0.180769 -0.118462  0.404615  
28      -0.057692  0.112308  -0.053846  0.143077  0.060769  0.192308

In [16]:
new_x = X_train[['goals_z', 'boxtouches_z', 'fouls_z', 'fouled_z', 'blocks_z']]

In [17]:
#add constant to predictor variables
new_x = sm.add_constant(new_x)

#fit regression model
model = sm.OLS(y_train, new_x).fit()

#view AIC of model
print(model.aic)

794.4357164871906


for AIC - the model with a lower AIC value is the better fitting model. 

https://www.statology.org/aic-in-python/

In [18]:
x = X_train[['goals_z', 'xg_z', 'tackles_z', 'nsxg_z']]

In [19]:
#add constant to predictor variables
x = sm.add_constant(x)

#fit regression model
model = sm.OLS(y_train, x).fit()

#view AIC of model
print(model.aic)

792.1161813534175


In [20]:
x_t = X_test[['goals_z', 'xg_z', 'tackles_z', 'nsxg_z']]

In [21]:
#add constant to predictor variables
x_t = sm.add_constant(x_t)

#fit regression model
model = sm.OLS(y_test, x_t).fit()

#view AIC of model
print(model.aic)

209.67105935243535


In [22]:
#add constant to predictor variables
x = sm.add_constant(X_test)

#fit regression model
model = sm.OLS(y_test, x).fit()

#view AIC of model
print(model.aic)

222.00398789642531


In [27]:
df.columns

Index(['goals_z', 'xg_z', 'crosses_z', 'boxtouches_z', 'passes_z',
       'progpasses_z', 'takeons_z', 'progruns_z', 'tackles_z',
       'interceptions_z', 'clearances_z', 'blocks_z', 'aerials_z', 'fouls_z',
       'fouled_z', 'nsxg_z', 'results'],
      dtype='object')

In [86]:
x_t = X_test[['goals_z', 'xg_z', 'nsxg_z','clearances_z', 
              'progpasses_z', 'interceptions_z']]

In [87]:
#add constant to predictor variables
x_t = sm.add_constant(x_t)

#fit regression model
model = sm.OLS(y_test, x_t).fit()

#view AIC of model
print(model.aic)

205.04062985068555


- <b>205.0406 = ['goals_z', 'xg_z', 'nsxg_z','clearances_z', 'progpasses_z', 'interceptions_z']</b>
- 209.6710 = ['goals_z', 'xg_z', 'tackles_z', 'nsxg_z']
- 222.0039 = all features

# Feature Selection

In [ ]:
from sklearn.feature_selection import SelectKBest, chi2


# https://scikit-learn.org/stable/modules/model_evaluation.html#regression-metrics

https://www.kaggle.com/code/satishgunjal/tutorial-k-fold-cross-validation